crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -769,10 +831,58 @@ where
                     }
 
                     self.visit_body(handler_body);
+
+                    // Each `except` block is mutually exclusive with all other `except` blocks.
+                    post_except_states.push(self.flow_snapshot());
+
+                    // It's unnecessary to do the `self.flow_restore()` call for the final except handler,
+                    // as we'll immediately call `self.flow_restore()` to a different state
+                    // as soon as this loop over the handlers terminates.
+                    if i < (num_handlers - 1) {
+                        self.flow_restore(pre_except_state.clone());
+                    }
                 }
 
+                // If we get to the `else` block, we know that 0 of the `except` blocks can have been executed,
+                // and the entire `try` block must have been executed:
+                self.flow_restore(post_try_block_state);
                 self.visit_body(orelse);
+
+                // Visiting the `finally` block!
+                //
+                // The state here could be that we visited exactly one `except` block,
+                // or that we visited the `else` block, **or** (unless the `except` branches
+                // are exhaustive) that we visited none of them!
+                // (An exception **could** have been raised that wasn't handled by any
+                // of the except handlers! But it would still be caught by the `finally`,
+                // which **always** runs.)
+                //
+                // N.B. Not all `try`/`except` clauses have `else` and/or `finally`!
+                // (In fact, some of them have `finally` but not `except`!)
+                // That's actually an irrelevant detail for us here, though, because
+                // a nonexistent `else` or `finally` block is just represented
+                // as an empty `Vec` in our AST, and visiting an empty `Vec` will not
+                // change the flow state at all.
+                for state in post_except_states {
+                    self.flow_merge(state);
+                }
+                if !exhaustive_except_branches {
+                    for state in visiting_try_block_states.into_snapshots() {
+                        self.flow_merge(state);
+                    }
+                }
+
                 self.visit_body(finalbody);
+
+                // Account for the fact that we could be visiting a nested `try` block,
+                // in which case the outer `try` block must be kept informed about all the possible
+                // between-definition states we could have encountered in the inner `try` block(!)
+                if let Some(current_try_block) =
+                    self.try_block_contexts().borrow_mut().current_try_block()
```

Comment:
Can't we call `self.try_block_contexts_mut()` here to avoid the `RefCell` in the contexts (and all the `borrow_mut` stuff)?